lxd/ubuntupro/client.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// GetGuestTokenV1DataAttributes returns a 403 Forbidden error if the host or the instance has GuestAttachSettingOff, otherwise
+// it calls the pro shim to get a token.
+func (s *Client) GetGuestTokenV1DataAttributes(ctx context.Context, instanceSetting GuestAttachSetting) (*ProAPIGetGuestTokenV1DataAttributes, error) {
+	if s.getGuestAttachSetting(instanceSetting) == GuestAttachSettingOff {
+		return nil, api.NewStatusError(http.StatusForbidden, "Guest attachment not allowed")
+	}
+
+	return s.pro.getProAPIGuestTokenDataAttributes(ctx)
+}
+
+// init configures the Client to watch the ubuntu advantage directory for file changes.
+func (s *Client) init(ctx context.Context, ubuntuAdvantageDir string, proShim pro) {
+	// Initial setting should be "off".
+	s.guestAttachSetting = GuestAttachSettingOff
+	s.pro = proShim
+
+	// Set up a watcher on the ubuntu advantage directory.
+	err := s.watch(ctx, ubuntuAdvantageDir)
+	if err != nil {
+		logger.Warn("Failed to configure Ubuntu configuration watcher", logger.Ctx{"err": err})
+	}
+}
+
+func (s *Client) watch(ctx context.Context, ubuntuAdvantageDir string) error {
+	// On first call, attempt to read the contents of the config file.
+	configFilePath := path.Join(ubuntuAdvantageDir, "interfaces", "lxd-config.json")
+	err := s.readConfigFile(configFilePath)
+	if err != nil && !os.IsNotExist(err) {
+		logger.Warn("Failed to read Ubunto Pro LXD configuration file", logger.Ctx{"err": err})
+	}
+
+	// Watch /var/lib/ubuntu-advantage for write, remove, and rename events.
+	monitor, err := drivers.Load(ctx, ubuntuAdvantageDir, fsmonitor.EventWrite, fsmonitor.EventRemove, fsmonitor.EventRename)
+	if err != nil {
+		return fmt.Errorf("Failed to create a file monitor: %w", err)
+	}
+
+	go func() {
+		// Wait for the context to be cancelled.
+		<-ctx.Done()
+
+		// On cancel, set the guestAttachSetting back to "off" and unwatch the file.
+		s.guestAttachSetting = GuestAttachSettingOff
+		err := monitor.Unwatch(path.Join(ubuntuAdvantageDir, "interfaces", "lxd-config.json"), "")
+		if err != nil {
+			logger.Warn("Failed to remove Ubuntu Pro configuration file watcher", logger.Ctx{"err": err})
+		}
+	}()
+
+	// Add a hook for the config file.
+	err = monitor.Watch(configFilePath, "", func(path string, event fsmonitor.Event) bool {
+		if event == fsmonitor.EventRemove {
+			// On remove, set guest attach to "off".
+			s.guestAttachSetting = GuestAttachSettingOff
+			return true
+		}
+
+		// Otherwise, read in the configuration.
+		err := s.readConfigFile(path)
+		if err != nil {
+			logger.Warn("Failed to read Ubunto Pro LXD configuration file", logger.Ctx{"err": err})
+		}
+
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to configure file monitor: %w", err)
+	}
+
+	s.monitor = monitor
+	return nil
+}
+
+// readConfigFile reads the Ubuntu Pro `lxd-config.json` file and sets appropriate values in the client.
+func (s *Client) readConfigFile(lxdConfigFile string) error {
```

Comment:
Can we call this `parseConfigFile` or `applyConfigFile` as "read" suggests a not mutating action, which this is not.